File: Easy/234_Palindrome_Linked_List/Solution.py
from typing import List, Optional
from devlibs.singly_linkedlist import ListNode


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head

        while (fast != None and fast.next != None):
            slow = slow.next
            fast = fast.next.next
        
        prev = None
        curr = slow
        
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        

        while (prev):
            if (head.val != prev.val):
                return False
            prev = prev.next
            head = head.next
        
        return True


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    head1 = ListNode(1, ListNode(2, ListNode(2, ListNode(1))))
    print(solution.isPalindrome(head1))  # Expected: True

    # test cases 2
    head2 = ListNode(1, ListNode(2))
    print(solution.isPalindrome(head2))  # Expected: False
    
    
    

File: Easy/234_Palindrome_Linked_List/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
*/

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* fast = head;
        ListNode* slow = head;

        while (fast!= nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* prev = nullptr;
        ListNode* curr = slow;

        while (curr != nullptr) {
            ListNode* next_node = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next_node;
        }

        while (prev != nullptr) {
            if (head->val != prev->val) {
                return false;
            }
            prev = prev->next;
            head = head->next;
        }
        
        return true;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;

    // test cases 1
    ListNode* head1 = createSinglyLinkList({1, 2, 2, 1});
    cout << solution.isPalindrome(head1) << endl;
    
    // test cases 2
    ListNode* head2 = createSinglyLinkList({1, 2});
    cout << solution.isPalindrome(head2) << endl;

    return 0;
}